### PR TITLE
[tt-train] Implement LoRA fine-tuning in Python

### PR DESCRIPTION
**Pull request by imichalakTT**
### Problem description

No Python-side support for LoRA.

### What's changed

- Added adapter.py with Adapter, IdentityAdapter, and ForwardInvocation — clean post-forward hook API
- Added lora.py with LoRA adapter and inject_lora — attaches low-rank adapters to target LinearLayer modules by name pattern
- Added train_lora_llama.py — minimal single-device Llama LoRA fine-tuning example on Shakespeare
- Fixed broken named_modules() and named_children() in ModuleList and ModuleDict

#### Why Adapter class instead of replacing/wrapping modules?

- Model structure stays the same, object references stay the same, property accesses also.
- When traversing the model you can still rely on `isinstance(mod, LinearLayer)` for other transformations, debugging, etc. No need for additional branches and special cases, e.g. `if isinstance(mod, LoRALinearLayer): mod = mod.base_layer`.
- Easily detach LoRA from existing LoRA augmented model just by detaching the adapter.
- Adapter approach is tiny, just mod.adapter = LoRA(), implementation too.
- Not a key feature, but you can compose adapters, for example, mod.adapter = ComposeAdapter(LoRA(), Other()).

#### FT script output

```
Total params: 31
Trainable (LoRA): 12
  * Llama/blocks/0/attention/kv_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/0/attention/kv_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/0/attention/kv_linear/weight: [1, 1, 128, 128]
  * Llama/blocks/0/attention/out_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/0/attention/out_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/0/attention/out_linear/weight: [1, 1, 128, 128]
  * Llama/blocks/0/attention/q_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/0/attention/q_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/0/attention/q_linear/weight: [1, 1, 128, 128]
    Llama/blocks/0/attention_norm/gamma: [1, 1, 1, 128]
    Llama/blocks/0/mlp/w1/weight: [1, 1, 512, 128]
    Llama/blocks/0/mlp/w2/weight: [1, 1, 128, 512]
    Llama/blocks/0/mlp/w3/weight: [1, 1, 512, 128]
    Llama/blocks/0/mlp_norm/gamma: [1, 1, 1, 128]
  * Llama/blocks/1/attention/kv_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/1/attention/kv_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/1/attention/kv_linear/weight: [1, 1, 128, 128]
  * Llama/blocks/1/attention/out_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/1/attention/out_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/1/attention/out_linear/weight: [1, 1, 128, 128]
  * Llama/blocks/1/attention/q_linear/LoRA/lora_A: [1, 1, 8, 128]
  * Llama/blocks/1/attention/q_linear/LoRA/lora_B: [1, 1, 128, 8]
    Llama/blocks/1/attention/q_linear/weight: [1, 1, 128, 128]
    Llama/blocks/1/attention_norm/gamma: [1, 1, 1, 128]
    Llama/blocks/1/mlp/w1/weight: [1, 1, 512, 128]
    Llama/blocks/1/mlp/w2/weight: [1, 1, 128, 512]
    Llama/blocks/1/mlp/w3/weight: [1, 1, 512, 128]
    Llama/blocks/1/mlp_norm/gamma: [1, 1, 1, 128]
    Llama/fc/weight: [1, 1, 96, 128]
    Llama/ln_fc/gamma: [1, 1, 1, 128]
    Llama/tok_emb/weight: [1, 1, 96, 128]
step    1/200  loss=4.5938
step   10/200  loss=4.5312
step   20/200  loss=4.4062
step   30/200  loss=4.2500
step   40/200  loss=4.1562
step   50/200  loss=4.1250
step   60/200  loss=4.1875
step   70/200  loss=4.0938
step   80/200  loss=4.1562
step   90/200  loss=4.0938
step  100/200  loss=4.0938
step  110/200  loss=4.1250
step  120/200  loss=4.1250
step  130/200  loss=4.1562
step  140/200  loss=4.0938
step  150/200  loss=4.0938
step  160/200  loss=4.1250
step  170/200  loss=4.0938
step  180/200  loss=4.1250
step  190/200  loss=4.0938
step  200/200  loss=4.0938
```

### Checklist

- [ ] APC (https://github.com/tenstorrent/tt-metal/actions/runs/22725433939)

### CODE REVIEW

File: tt-train/sources/ttml/ttml/modules/module_base.py [hunk 1]
```diff
@@ -4,61 +4,80 @@
...
-        # Auto-register tensors
-        tensor = None
-        if isinstance(value, Parameter):
-            tensor = value.tensor
-        elif isinstance(value, Buffer):
-            self._buffers[name] = value.tensor
+        if isinstance(attr_value, Adapter):
+            adapter_name = type(attr_value).__name__
+            for f_name, f_value in attr_value.parameters().items():
+                self._bind_tensor(f_value.tensor, f"{adapter_name}/{f_name}")
```
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/modules/module_base.py:42]: will address this

File: tt-train/sources/examples/lora_llama/train_lora_llama.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        num_attention_heads=4,
+        num_key_value_heads=2,
+        vocab_size=vocab_size,
+        max_position_embeddings=SEQ_LEN,
+    )
+)
+
+model = inject_lora(
+    model, rank=LORA_RANK, alpha=LORA_ALPHA, target_modules=LORA_TARGET_MODULES
+)
```
Comment (philei-tt) [inline tt-train/sources/examples/lora_llama/train_lora_llama.py:64]: imo having LoraConfig similar to the one in [PEFT](https://huggingface.co/docs/diffusers/en/training/lora) would be a bit nicer. Plus you can add a static method to create it from yaml.

File: tt-train/sources/ttml/ttml/modules/lora.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+class LoRA(Adapter):
+    """Low-Rank Adaptation: adds lora_B(lora_A(x)) * scaling to the base output."""
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        rank: int,
+        alpha: float,
+        use_rslora: bool = False,
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/modules/lora.py:46]: I think you should add `is_bias_trainable` (unfreeze original bias tensor) and `dropout` parameters

File: tt-train/sources/ttml/ttml/modules/module_base.py [hunk 4]
```diff
@@ -4,61 +4,80 @@
...
+            self._bind_module(attr_value, attr_name)
             return
 
-        # Auto-register tensors
-        tensor = None
-        if isinstance(value, Parameter):
-            tensor = value.tensor
-        elif isinstance(value, Buffer):
-            self._buffers[name] = value.tensor
+        if isinstance(attr_value, Adapter):
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttml/modules/module_base.py:39]: I'm not a fan of introducing Adapter class + making module base depend on it. I believe in PEFT they make LoraLinear just another module, and when injecting lora, they just store original linear inside LoraLinear, and override original linear module with LoraLinear. this way you don't need to make module base aware of adapters.

**Review by philei-tt — CHANGES_REQUESTED** [review]

File: tt-train/sources/examples/lora_llama/train_lora_llama.py (hunk 2 above)
Comment (imichalakTT) [reply tt-train/sources/examples/lora_llama/train_lora_llama.py:64]: Absolutely, I will work on a higher-level API for this, closer to PEFT. inject_lora plays a role of a low-level API

File: tt-train/sources/ttml/ttml/modules/lora.py (hunk 3 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/modules/lora.py:46]: Yes, though I think we should put it on a LoRA config and implement freezing and dropout inside the PEFT-style helper.

File: tt-train/sources/ttml/ttml/modules/module_base.py (hunk 4 above)
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/modules/module_base.py:39]: By making this design choice I decided that keeping model stable is worth the trade-offs. Alternative solutions: 
a) replace LinearLayer with LoRALinearLayer (weight, bias, lora_A, lora_B flat on the module), we destroy LinearLayer, but accessing .bias or .weight works the same.
b) we introduce LoRALinearLayer that wraps existing LinearLayer, we can add .bias and .weight that return .base_layer.bias and .base_layer.weight. probably better overall than a), but we still change the structure.

Modifying the structure is a viable way, but LoRALinearLayer stops being just LinearLayer, now unless it's a subclass we have to account in our code that we need special handling for it, because isinstance(mod, LinearLayer) stops counting LoRA extended layers. Further extension is also harder, with adapters we can just do module.adapter = ComposedAdapter(LoRA, ...). Also debugging, measuring, temporarily disabling LoRA, I think can be easier with adapters.

File: tt-train/sources/ttml/ttml/modules/module_base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/ttml/ttml/modules/module_base.py:42]: unfortunately C++ doesn't have tensor unregistration. i think it's best to address this when migrating to fully Python module class

File: tt-train/sources/examples/lora_llama/train_lora_llama.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+LORA_ALPHA = 16
+LORA_TARGET_MODULES = ["q_linear", "kv_linear", "out_linear"]
+
+# ── Data ──────────────────────────────────────────────────────────────────────
+
+set_seed(42)
+
+text = load_shakespeare_text()
+tokenizer = CharTokenizer(text)
+vocab_size = (tokenizer.vocab_size + 31) // 32 * 32
```
Comment (mdragulaTT) [inline tt-train/sources/examples/lora_llama/train_lora_llama.py:39]: Where does those constants come from?

File: tt-train/sources/ttml/ttml/modules/adapter.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Mapping, Sequence
+
+
+@dataclass(frozen=True)
+class ForwardInvocation:
```
Comment (mdragulaTT) [inline tt-train/sources/ttml/ttml/modules/adapter.py:12]: Maybe it’s a dumb question, but it’s not clear to me what this class is for. Could you extend the PR description or add a brief comment here explaining why the forward is encapsulated in a separate class? 🙈

File: tt-train/sources/ttml/ttml/modules/adapter.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    def __call__(self, fwd: ForwardInvocation) -> Any:
+        """Return the forward output unchanged."""
+        return fwd.output
+
+    def parameters(self) -> dict:
+        """Return trainable parameters keyed by name. Override in subclasses."""
+        return {}
+
+
+class IdentityAdapter(Adapter):
```
Comment (mdragulaTT) [inline tt-train/sources/ttml/ttml/modules/adapter.py:46]: Why do you need no-op adapter?

File: tt-train/sources/ttml/ttml/modules/lora.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+import ttnn
+import ttml
+
+from .adapter import Adapter, ForwardInvocation
+from .linear import LinearLayer
+from .parameter import Parameter
+
+
+def _create_lora_A(in_features: int, rank: int):
```
Comment (mdragulaTT) [inline tt-train/sources/ttml/ttml/modules/lora.py:20]: Could you add the return types for those functions?

**Review by mdragulaTT — COMMENTED** [review]
I’ve lost the thread a bit with LoRA, so maybe this is a dumb question, but why isn’t the [LoRA C++ code merged yet](https://github.com/tenstorrent/tt-metal/pull/34434)? Wouldn’t it be easier to add the bindings on top of the merged C++ implementation so we can run it from Python?
